ksql-engine/src/main/java/io/confluent/ksql/connect/ConnectPollingService.java

```diff
@@ -0,0 +1,158 @@
+ * and for each one it will poll kafka and schema registry to see if any topics
+ * created by that connector are eligible for automatic registry with KSQL on
+ * a regular basis.
+ *
+ * <p>Ideally, Connect would implement a metadata topic where it publishes all
+ * this information on a push basis, so we don't need to poll kafka or schema
+ * registry.</p>
+ */
+@ThreadSafe
+class ConnectPollingService extends AbstractScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ConnectPollingService.class);
+  private static final int INTERVAL_S = 30;
+
+  private final KsqlExecutionContext executionContext;
+  private final Consumer<CreateSource> sourceCallback;
+
+  private Set<Connector> connectors;
+
+  ConnectPollingService(
+      final KsqlExecutionContext executionContext,
+      final Consumer<CreateSource> sourceCallback
+  ) {
+    this.executionContext = Objects.requireNonNull(executionContext, "executionContext");
+    this.sourceCallback = Objects.requireNonNull(sourceCallback, "sourceCallback");
+    this.connectors = ConcurrentHashMap.newKeySet();
+  }
+
+  /**
+   * Add this connector to the set of connectors that are polled by this
+   * {@code ConnectPollingService}. Next time an iteration is scheduled in
+   * {@value #INTERVAL_S} seconds, the connector will be included in the topic
+   * scan.
+   *
+   * @param connector a connector to register
+   */
+  void addConnector(final Connector connector) {
+    connectors.add(connector);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    if (connectors.isEmpty()) {
+      // avoid making external calls if unnecessary
+      return;
+    }
+
+    try {
+      final Set<String> topics = executionContext.getServiceContext()
+          .getAdminClient()
+          .listTopics()
+          .names()
+          .get(10, TimeUnit.SECONDS);
+
+      final Set<String> subjects = ImmutableSet.copyOf(
+          executionContext.getServiceContext()
+              .getSchemaRegistryClient()
+              .getAllSubjects()
+      );
+
+      for (final String topic : topics) {
+        final Optional<Connector> maybeConnector = connectors
+            .stream()
+            .filter(candidate -> candidate.matches(topic))
+            .findFirst();
+        maybeConnector.ifPresent(connector -> handleTopic(topic, subjects, connector));
+      }
+    } catch (final Exception e) {
+      LOG.error("Could not resolve connect sources. Trying again in {} seconds.", INTERVAL_S, e);
+    }
+  }
+
+  private void handleTopic(
+      final String topic,
+      final Set<String> subjects,
+      final Connector connector
+  ) {
+    final String valueSubject = topic + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
+    if (subjects.contains(valueSubject)) {
+      final String name = connector.getName();
+      final String source = connector.mapToSource(topic).toUpperCase();
+
+      // if the meta store already contains the source, don't send the extra command
```

Comment:
that might be a way to go about it, but that requires adding an index in the Metastore. For now, I just want to make sure I'm not spamming the command topic with unnecessary commands and this serves that purpose. We can always change that later, it would be mostly backwards compatible